---
title: "Analyse statistique des donn&eacute;es g&eacute;nomiques (ASG)"
author: "Denis Puthier & Jacques van Helden"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    fig_captions: yes
  md_document:
    variant: markdown_github
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 2
  word_document:
    toc: no
    toc_depth: 2
#css: ../../html/course.css
---

## Web site for this course

<https://dputhier.github.io/ASG/>

## Teachers

| Abbrev | Name  |
|----------|--------------------------------------------------|
| DP |  Denis Puthier |
| JvH | [Jacques van Helden](http://jacques.van-helden.perso.luminy.univ-amu.fr/) |
| | |

## Year

2017-2018

## Audience

- 2ème année du Master en Bioinformatique, biochimie structurale et génomique (BBSG).
- Ecole doctorale


<!--
## Travail personnel / Personal work

The instructions for the 2016 personal work can be found here

-  [html document (human-readable)](practicals/rnaseq_psoriasis_Li2014/RNA-seq_psoriasis_analysis_exercises.html)

- [Rmd (including code to load the data)](practicals/rnaseq_psoriasis_Li2014/RNA-seq_psoriasis_analysis_exercises.Rmd)

-->

## Resources

| Tool | About |
|---------------------------------------|---------------------------------------------------------------|
|[R](http://www.r-project.org/) |	A free software environment for statistical computing and graphics |
|[R markdown](http://rmarkdown.rstudio.com/) |	Documentation for R markdown language (used for the practicals) |
|[Bioconductor](http://www.bioconductor.org) |	A set of R libraries dedicated to statistical analysis of genomics data.|
| [MeV: MultiExperiment Viewer](http://www.tm4.org/) |	A Java application designed to allow the analysis of expression data |
|[Cluster 3.0](http://bonsai.hgc.jp/~mdehoon/software/cluster/software.htm) | Implements the most commonly used clustering methods for gene expression |
| [java Treeview](http://jtreeview.sourceforge.net) |	Java-based tool to visualize trees prodced by hierarchical clustering togeter with a heatmap with expression proviles.|

## Prerequisites

Students are expected to have followed the introduction to statistics in the first year of the master. 

- Probabilités et statistiques pour la biologie (SBBAU16L - STAT1): <https://jvanheld.github.io/stat1/>

We assume that the following concepts are acquired. 

- Discrete distributions (geometric, binomial)
- Sampling and estimation
- Mean comparison tests (Student, Welch)

A basic knowledge of the R language is expected. 

- handling of variables and data frames ("tables")
- distributions of probability
- drawing (histograms, dot plots)
- test of hypothesis

If you did not receive any training.

During the first practical we will briefly revise these concepts and practical skills. 

## Schedule and contents

| Date   | From | To   | Subject         | Teacher | Concepts   | Material            |
|-------|-----|-----|----------------|-----|----------------|-------------------------------------|
| 
| 3/11 | 14:00 | 18:00 | Introduction au cours | JvH | | Diapos [html](slides/01_introduction.html) [Rmd](slides/01_introduction.Rmd) |
| 3/11<br>6/11 | 14:00 | 18:00 | Detecting differentially expressed genes (DEG) with microrarrays | DP |- Hypothesis testing<br>- Student $t$ statistics<br>- Unbiased estimation of variance<br>- MA plot<br>- Volcano plots<br> -P-value distribution <br>- E-value| - **Slides:** [html](slides/Den_Boer_use_case/Den_Boer_use_case.html) [Rmd](slides/Den_Boer_use_case/Den_Boer_use_case.Rmd) <br> - **Basics about Student and Welch’s t test** [html](practicals/microarrays_student_test/Student_test_info.html) [Rmd](practicals/microarrays_student_test/Student_test_info.Rmd) <br> -  **Practical:** detecting differentially expressed genes in microarray data [html](practicals/microarrays_student_test/DenBoer_Student_test.html) [Rmd](practicals/microarrays_student_test/DenBoer_Student_test.Rmd) <br>  -  **Practical:** generating random control sets following a Normal distribution (incomplete) [html](practicals/microarrays_student_test/negative_control_Welch-test.Rmd) [Rmd](practicals/microarrays_student_test/negative_control_Welch-test.Rmd)|
| 7/11 | 14:00  | 18:00 | About distances and clustering | DP | - Distance metrics <br> - Hierarchical clustering <br> | - Theory : [html](practicals/unsup_classif_slides/unsup_classification.pdf) <br>- Distance metrics and clustering (practical) [html](practicals/distances/distances.html) [rmd](practicals/distances/distances.Rmd)<br>|
| 20/11 | 9:00 | 12:30 | The multiple ways to correct for multiple testing | JvH | - False positive risk (FPR) <br> - Expected number of false positives (E-value) <br> - Family-Wise Error Rate (FWER) <br> - False Discovery Rate (FDR) | - [Multiple testing corrections  (slides)](http://pedagogix-tagc.univ-mrs.fr/courses/ASG1/pdf_files/06.04.multitesting_corrections.pdf) <br> - [Multiple testing corrections (practical)](practicals/multiple_testing_td/multiple_testing_corrections.html) |
| 20/11 | 9:00 | 12:30 | Supervised classification  | JvH | - Discriminant analysis <br>- Cross-validation (k-fold, LOO) <br>- Data dimensionality and overfitting <br>- Variable selection | - [Introduction to multivariate analysis](pdf_files/07.00.multiv_intro.pdf) <br> - [Discriminant analysis (slides)](pdf_files/11.discriminant_analysis.pdf)<br>- [Dimension reduction and PCA](pdf_files/10.visualization.pdf)<br>- [Practical: supervised classification](practicals/supervised_classification_td/supervised_classification_td.html)  |
| 21/11 | 14:00  | 18:00 | Functional enrichment of DEG | DP | - Functional enrichment statistics <br> - The hypergeometric distribution <br> | - Theory : on white board <br> - [Hypergeometric distribution and enrichment statistics. An example application: DAVID (practical)](practicals/go_statistics_td/go_statistics_td_2015.html)<br>|
| 22/11 | 14:00  | 18:00  | Overview of discrete distributions, with applications to NGS | DP | - Geometric<br>- Binomial<br>- Poisson<br>-Hypergeometric<br>- Negative binomial | Tutorial [[html](practicals/discrete_distributions_overview/discrete_distributions.html)][[pdf](practicals/discrete_distributions_overview/discrete_distributions.pdf)][[Rmd](practicals/discrete_distributions_overview/discrete_distributions.Rmd)] |
| 22/11  | 14:00 | 18:00 | Detecting differentially expressed genes (DEG) with RNA-seq | DP/JvH | - RNA-Seq principles <br>- Normalizing RNA-seq counts <br>-Detecting differentially expressed genes (DEG)| - RNA-Seq DEG with DESeq2 [[html]](practicals/rnaseq_diff_Snf2/rnaseq_diff_Snf2.html) [[pdf]](practicals/rnaseq_diff_Snf2/rnaseq_diff_Snf2.pdf) [[Rmd]](practicals/rnaseq_diff_Snf2/rnaseq_diff_Snf2.Rmd) |
| 22/11  | 14:00 | 18:00 | Descriptive statistics with ggplot2 | DP | - ggplot2 principles <br>- Layout, creating diagrams | - Introduction to ggplot2 [[html]](practicals/intro_ggplot/intro_ggplot.html) |
|  |  |  | RNA-seq analysis (pursued) | JvH | | [[]]()<br>[[]]() |
|  |  |  | Preparation of the personal work | DP | **Homework: analysis of psoriasis data** [[html]](practicals/rnaseq_psoriasis_Li2014/RNA-seq_psoriasis_analysis_exercises.html) [[Rmd]](practicals/rnaseq_psoriasis_Li2014/RNA-seq_psoriasis_analysis_exercises.Rmd) [[bib]](practicals/rnaseq_psoriasis_Li2014/RNA-seq_psoriasis_analysis_biblio.bib)|

## Screen pictures

[![Multivariate data](screen_pictures_2017/00_multivariate_data.jpg)](screen_pictures_2017/00_multivariate_data.jpg)

[![Approches de classification (supervisée ou non). ](screen_pictures_2017/01-02_classification.jpg)](screen_pictures_2017/01-02_classification.jpg)

[![3. Réduction de dimensionnalité. 4. Evaluation des résultats de classification supervisée ](screen_pictures_2017/03-04_dimensionalite-evaluation.jpg)](screen_pictures_2017/03-04_dimensionalite-evaluation.jpg)

[![5. Quelques méthodes de classification supervisée. ](screen_pictures_2017/05_illlus-classifieurs.jpg)](screen_pictures_2017/05_illlus-classifieurs.jpg)

## Additional support

| Concept | Description |
|----------------|------------------------------------------------|
|Introduction to R | - [First steps with R and Siméon Denis Poisson  (practical)](http://jvanheld.github.io/EBA15_stats_RStudio/practicals/01_first_steps_with_R/01_first_steps_with_R.html)<br> - [R language: A quick tutorial (practical)](practicals/r_intro_td/index.html)<br> - [R language: A quick tutorial (practical)](practicals/r_intro_en/RQuickTutorial.pdf)|
|Occurrence statistics| - [The Poisson distribution in the context of Peak-calling (practical)](http://jvanheld.github.io/EBA15_stats_RStudio/practicals/02_peak-calling/peak-calling_stats_practical.html)<br>- [The Poisson distribution in the context of k-mers occurence statistics (practical)](http://dputhier.github.io/STAT1_R/practical/R_intro.html)<br> - [Read mapping statistics and the binomial distribution (practical)](practicals/chip-seq_read_mapping/read_mapping_solutions.html)<br> - [Hypergeometric distribution and enrichment statistics. An example application: DAVID (practical)](practicals/go_statistics_td/go_statistics_td_2015.html) <br> - [Application example: K-mer occurrences in ChIP-seq peaks (practical)](http://pedagogix-tagc.univ-mrs.fr/courses/ASG1/practicals/kmer_occurrences/kmer_occurrences_chip-seq_peaks.html)|
|Microarray analysis | - [Introduction to multivariate analysis (slides)](pdf_files/07.00.multiv_intro.pdf)<br> - [Transcriptome microarrays: study cases (slides)](pdf_files/07.01.multiv_study_cases.pdf)<br> - [Normalization of Affymetrix DNA chip (slides)](practicals/affynorm_slides/index.html) - <br> - [Handling and normalizing affymetrix data with bioconductor (practical)](practicals/affynorm_td/index.html) <br> - [Differential_expression (slides)](pdf_files/07.04.differential_expression.pdf) <br> - [Basics about Student and Welch’s t test](practicals/microarrays_student_test/Student_test_info.html) <br> - [Microarray data: selecting differentially expressed genes with R or TmeV (practical)](practicals/affy_selection_td/affy_selection_td.html) <br> - [Sampling distributions (practical)](practicals/sampling_distrib.html) - [Detecting differentially expressed genes in microarray data. Part I: exploring Student t statistics](practicals/microarrays_student_test/DenBoer_Student_test.html)|
|The multiple ways to correct multiple testing| - [Multiple testing corrections  (slides)](http://pedagogix-tagc.univ-mrs.fr/courses/ASG1/pdf_files/06.04.multitesting_corrections.pdf) <br> - [Multiple testing corrections (practical)](practicals/multiple_testing_td/multiple_testing_corrections.html)|
|Clustering (unsupervised classification)| - [Correlation analysis (slides)](pdf_files/07.03.correlation_analysis.pdf) <br> - [Clustering (slides)](pdf_files/09.clustering.pdf) <br> - [Clustering (slides DP)](practicals/unsup_classif_slides/unsup_classification.pdf)<br> - [About distances](practicals/distances/distances.html) <br> - [Handling clustering methods: artificial datasets (practical)](unsup_classification_sim_td/index.html) <br> - [Clustering of microarray data](practicals/unsup_classification_td/index.html)|
|RNA-seq data analysis| - [RNA-Seq method (slides)](pdf_files/Rnaseq_2016.pdf) <br> - [RNA-seq read mapping (practical)](practicals/rnaseq_td/rnaseq_td.html)<br> - [The negative binomial and DESeq bioC package](practicals/rnaseq_diff/DESeqTuto.html) <br> - [The negative binomial and DESeq bioC package](practicals/rnaseq_diff/rnaseq_diff.html)|
|Supervised classification| - [Introduction to multivariate analysis](pdf_files/07.00.multiv_intro.pdf) <br> - [Discriminant analysis (slides)](pdf_files/11.discriminant_analysis.pdf)|
|Visualization| - [Dimension reduction and PCA](pdf_files/10.visualization.pdf)|
| Overview | - [Discrete distributions for NGS data analysis](practicals/discrete_distributions_overview/discrete_distributions.html)

## Personal work

The instructions for the 2017 personal work can be found here

-  [html document (human-readable)](practicals/recount/RNA-seq_recount_analysis_exercises.html)

- [Rmd](practicals/recount/RNA-seq_recount_analysis_exercises.Rmd)

