---
title: 'RNA-seq: analysis of Psoriasis data (Li et al, 2014)'
author: "Denis Puthier and Jacques van Helden"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  slidy_presentation:
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    highlight: tango
    incremental: no
    keep_md: no
    self_contained: yes
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_md: no
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
bibliography: "RNA-seq_psoriasis_analysis_biblio.bib"
#css: ../../html/course.css
---




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, cache = TRUE, message = FALSE, warning = FALSE,
                      comment = "")

## Options to display pretty numbers
library(knitr)
knit_hooks$set(inline = function(x) {
  prettyNum(x, big.mark=" ")
})
options(scipen = 6, digits = 3)

```

# Exercise


Li and coworkers [-@Li:2014dp] used RNA-seq to characterize the transcriptome of 92 skin samples of people suffering from psoriasis, and 82 control samples. We pre-processed the raw sequences (NGS reads) to obtain a data table containing the counts per gene (row) for each sample (columns).  


## Goals of this practical

1. **Detection and interpretation of differentially expressed genes:** run a statistical analysis of the read counts, in order to detect differentially expressed genes (**DEG**) between psoriasis and control samples [@Li:2014dp], and to study the functional associations of these genes, and provide a biological interpretation of these results. 

2. **Methodological evaluation:** 

    a. **Robusntess analysis:** analyse the impact of individual particularities of the patients/controls, and measure the impact of the sample size by applying a sub-sampling approach. 

    b. **Negative control:** check the reliability of the DEG detection method (DESeq2) by running the same analysis with two subset of samples belonging to the same group. 


****************************************************************
## Dataset

The raw datasets were downloaded from Gene Expression Omnibus series **[GSE54456](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE54456)**.

### Processing

(**<font color="red">Denis, please give the parameters</font>**)


### Access to the count tbale

Count table: [data/counts_all_sample_header.txt.gz](data/counts_all_sample_header.txt.gz)

```{r}
dir.psoriasis <- ("~/ASG/practicals/rnaseq_psoriasis_Li2014/")
dir.counts <- file.path(dir.psoriasis, "data")
file.counts <- file.path(dir.counts, "counts_all_sample_header.txt.gz")

## Create a directory to download the dataset if it does not exist
dir.create(dir.counts, showWarnings = FALSE, recursive = TRUE)


## Download the psoriasis file if required
if (!file.exists(file.counts)) {
  url.counts <- "https://dputhier.github.io/ASG/practicals/rnaseq_psoriasis_Li2014/data/counts_all_sample_header.txt.gz"
  download.file(url=url.counts, destfile = file.counts)
}

counts <- read.delim(file.counts, row.names = 1)
names(counts) <- sub(pattern="_count.txt", replacement="", x=names(counts))  
names(counts) <- sub(pattern="__", replacement="_", x=names(counts))  
# View(counts)
head(names(counts))
```

The count table contains `r nrow(counts)` rows and `r ncol(counts)` columns.

```{r pheno_table}
## Generate a "pheno" table (description of each sample) from the headers of the count table. This is a bit tricky, but since we are often led to use such tricks it is good to see how to proceed.


## Load the limma library which implement strplit2, muech more convenient than strsplit
if (!require(limma)) {
  source("http://bioconductor.org/biocLite.R")
  biocLite("limma")  
}
library(limma)

psoriasis.pheno <- data.frame(strsplit2(names(counts), split = "_"))
class(psoriasis.pheno)
head(psoriasis.pheno)
names(psoriasis.pheno) <- c("GSM_ID", "M_ID", "Group","tissue", "SRX_ID")
row.names(psoriasis.pheno)<- names(counts)
# head(psoriasis.pheno)
# View(psoriasis.pheno)
# View(counts)

## Write a file with the pheno
write.table(psoriasis.pheno, file=file.path(dir.counts, "counts_all_sample_pheno.tab"), quote=FALSE, row.names = TRUE, col.names = TRUE, sep="\t")

```


****************************************************************
# References



