---
title: 'RNA-seq: analysis of Psoriasis data (Li et al, 2014)'
author: "Denis Puthier and Jacques van Helden"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  slidy_presentation:
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    highlight: tango
    incremental: no
    keep_md: no
    self_contained: yes
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_md: no
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
bibliography: "RNA-seq_psoriasis_analysis_biblio.bib"
#css: ../../html/course.css
---




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, cache = TRUE, message = FALSE, warning = FALSE,
                      comment = "")

## Options to display pretty numbers
library(knitr)
knit_hooks$set(inline = function(x) {
  prettyNum(x, big.mark=" ")
})
options(scipen = 6, digits = 3)

```

# Exercise


Li and coworkers [-@Li:2014dp] used RNA-seq to characterize the transcriptome of 92 skin samples of people suffering from psoriasis, and 82 control samples. We pre-processed the raw sequences (NGS reads) to obtain a data table containing the counts per gene (row) for each sample (columns).  


## Goals of this practical

1. **Detection and interpretation of differentially expressed genes:** run a statistical analysis of the read counts, in order to detect differentially expressed genes (**DEG**) between psoriasis and control samples [@Li:2014dp], and to study the functional associations of these genes, and provide a biological interpretation of these results. 

2. **Methodological evaluation:** 

    a. **Robusntess analysis:** analyse the impact of individual particularities of the patients/controls, and measure the impact of the sample size by applying a sub-sampling approach. 

    b. **Negative control:** check the reliability of the DEG detection method (DESeq2) by running the same analysis with two subset of samples belonging to the same group. 


****************************************************************
## Dataset

The raw datasets were downloaded from Gene Expression Omnibus series **[GSE54456](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE54456)**.

### Processing

(**<font color="red">Denis, please give the parameters</font>**)


### Access to the count tbale


```{r}
## If required, download the count table from the Web site (this is done only once).
url.counts <- "https://dputhier.github.io/ASG/practicals/rnaseq_psoriasis_Li2014/data/counts_all_sample_header.txt.gz"

## Local paths
dir.psoriasis <- ("~/ASG/practicals/rnaseq_psoriasis_Li2014/")
dir.counts <- file.path(dir.psoriasis, "data")
file.counts <- file.path("data", "counts_all_sample_header.txt.gz")

## Create a directory to download the dataset if it does not exist
dir.create(dir.counts, showWarnings = FALSE, recursive = TRUE)
setwd(dir.psoriasis)

## Download the psoriasis file if required
if (!file.exists(file.counts)) {
  download.file(url=url.counts, destfile = file.counts)
}

counts <- read.delim(file.counts, row.names = 1)
names(counts) <- sub(pattern="_count.txt", replacement="", x=names(counts))  
names(counts) <- sub(pattern="__", replacement="_", x=names(counts))  
# View(counts)
# head(names(counts))
kable(counts[101:108, 1:3], caption = "**Table 1.** Small piece of the count table. ")
```

The [count table](`r file.counts`) contains `r nrow(counts)` rows (one per gene) and `r ncol(counts)` columns (one per sample).

```{r pheno_table}
## Generate a "pheno" table (description of each sample) from the headers of the count table. This is a bit tricky, but since we are often led to use such tricks it is good to see how to proceed.


## Load the limma library which implement strplit2, muech more convenient than strsplit
if (!require(limma)) {
  source("http://bioconductor.org/biocLite.R")
  biocLite("limma")  
}
library(limma)

psoriasis.pheno <- data.frame(strsplit2(names(counts), split = "_"))
names(psoriasis.pheno) <- c("GSM_ID", "M_ID", "Group","tissue", "SRX_ID")
row.names(psoriasis.pheno)<- names(counts)
# head(psoriasis.pheno)
# View(psoriasis.pheno)

## Write a file with the pheno
setwd(dir.psoriasis)
file.pheno <- file.path("data", "counts_all_sample_pheno.tab")
write.table(psoriasis.pheno, file=file.pheno, 
            quote=FALSE, row.names = TRUE, col.names = TRUE, sep="\t")

```

The [pheno table](`r file.pheno`) contains `r nrow(psoriasis.pheno)` rows (one per sample) and `r ncol(psoriasis.pheno)` columns (one per sample attribute). 

```{r}
kable(head(psoriasis.pheno), row.names = FALSE, caption = "**Table 2. First rows of the pheno table. ** Each row describes one sample. ")

kable(as.data.frame(table(psoriasis.pheno$Group)), col.names = c("Group", "Number of samples"),
      caption = "**Table 3. Number of samples per group.** ")
```

## Analyses

1. **Differential analysis**. Analyse the full dataset to detect differentially expressed genes. This analysis should include the following steps.

    a. Log2-transformation of the counts (with an epsilon). 
    b. Graphical description of the data (historgrams, barplots, boxplots, ...).
    c. Computation of summary statistics per sample (min, max, mean, median, quartiles, number of zeros, ...) for raw counts and log2-transformed counts.
    d. Detection of differentially expressed genes (including graphical representations: volcano plot, p-value histogram, ...). 
    e. Functional enrichment of the differentially expressed genes. 

2. **Sub-sampling**. Run the same analysis on randomly selected subsets of the samples, with various subset sizes (n=2,3,4,5,10,20,40). 

3. **Negative control. ** Run the same analysis with two subsets of samples belonging to the same group (psoriasis versus psoriasis, control versus control).

## Reports

Les rapports peuvent être rédigés en français ou en anglais.

**Format:** the report can be submitted in either html or pdf format. The original Rmd file that was used to generate the report must be submitted together with the report. This file should allow the teachers to reprodue the analysis on their computers. 

****************************************************************
# References



