<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <link rel="stylesheet" type="text/css" href="../../html/course.css" media="screen,projection" />
    <title>ChIP-Seq: quality filtering, read mapping and peak-calling</title>
    <script type="text/javascript" src="../syntaxhighlight/shCore.js"></script>
    <script type="text/javascript" src="../syntaxhighlight/shBrushBash.js"></script>
    <link type="text/css" rel="stylesheet" href="../syntaxhighlight/shCore.css"/>
    <link type="text/css" rel="stylesheet" href="../syntaxhighlight/shThemeDefault.css"/>
    <script type="text/javascript">
      SyntaxHighlighter.config.clipboardSwf = '../../style/syntaxhighlight/clipboard.swf';
      SyntaxHighlighter.all();
    </script>
    
  </head>
  <body>
    
    <script language="JavaScript" type="text/javascript">
      <!--
	  function sizeTbl2(h,i) {
	  var tbl = document.getElementById(i);
	  tbl.style.display = h;
	  }
	  // -->
    </script> 


<h1>ChIP-Seq: quality filtering, read mapping and peak-calling</h1>
  
      <a name='toc'></a><h2>Contents</h2>

  
    <ul>
	<li><a href='#intro'>CHiP-Seq dataset description</a></li>
<!--	<li><a href='#flowchart'>Flow chart of the analysis</a></li> -->
	<li><a href='#geo'>Getting informations about the experiment using the GEO and SRA websites</a></li>
	<li><a href='#qual'>Quality control of sequencing data</a></li>
	<li><a href='#trim'>Read trimming and filtering</a></li>
	<li><a href='#bowtie'>Mapping reads with bowtie</a></li>
	<li><a href='#igv'>Visualizing the results with Integrated Genome Browser (IGV).</a></li>
	<li><a href='#input'>Mapping reads with bowtie (input)</a></li>	
	<li><a href='#peaks'>Peak calling</a></li>	
<!--	<li><a href='#troubles'>If you get troubles</a>-->
    </ul>

<hr><a name='intro'></a><h2>CHiP-Seq dataset description</h2>
      
<p>For this tutorial we will use CHiP-Seq datasets produced by <a href='http://www.ncbi.nlm.nih.gov/pubmed/23172872'>Theodorou <i>et al</i></a>. The authors used ChIP-Seq technology in order to systematically identify ESR1 binding regions across the human genome. The corresponding abstract of this article is provided below.</p>

<h4>Abstract</h4>
<p>Estrogen receptor (ESR1) drives growth in the majority of human breast cancers by binding to regulatory elements and inducing transcription events that promote tumor growth. Differences in enhancer occupancy by ESR1 contribute to the diverse expression profiles and clinical outcome observed in breast cancer patients. GATA3 is an ESR1-cooperating transcription factor mutated in breast tumors; however, its genomic properties are not fully defined. </p>
<p>In order to investigate the composition of enhancers involved in estrogen-induced transcription and the potential role of GATA3, we performed extensive ChIP-sequencing in unstimulated breast cancer cells and following estrogen treatment. We find that GATA3 is pivotal in mediating enhancer accessibility at regulatory regions involved in ESR1-mediated transcription. GATA3 silencing resulted in a global redistribution of cofactors and active histone marks prior to estrogen stimulation. These global genomic changes altered the ESR1-binding profile that subsequently occurred following estrogen, with events exhibiting both loss and gain in binding affinity, implying a GATA3-mediated redistribution of ESR1 binding. The GATA3-mediated redistributed ESR1 profile correlated with changes in gene expression, suggestive of its functionality. Chromatin loops at the TFF locus involving ESR1-bound enhancers occurred independently of ESR1 when GATA3 was silenced, indicating that GATA3, when present on the chromatin, may serve as a licensing factor for estrogen-ESR1-mediated interactions between cis-regulatory elements. Together, these experiments suggest that GATA3 directly impacts ESR1 enhancer accessibility, and may potentially explain the contribution of mutant-GATA3 in the heterogeneity of ESR1+ breast cancer.</p>

<!--<hr><a name='flowchart'></a><h2>Flow chart</h2>
   
<p align="center"><a target='main' href='img/flow_chart.png'><img style="text-align:center" border='1' src='img/flow_chart.png'></a></p> -->
	
<hr><a name='geo'>
</a><h2>Getting informations about the experiment using the GEO and SRA websites</h2>

<p>Gene Expression Omnibus (<a target="_blank" href='http://www.ncbi.nlm.nih.gov/geo/'>GEO</a>) is a public repository that provide tools to submit, access and mine functional genomics data.  Data may be related to array- or sequence-based technologies. For HTS data, GEO provides both processed data (such as *.bam, *.bed, *.wig files) and links to raw data. Raw data are available from the Sequence Read Archive (<a target="_blank" href='http://trace.ncbi.nlm.nih.gov/Traces/sra/'>SRA</a>) database (including 454, IonTorrent, Illumina, SOLiD, Helicos and Complete Genomics). Both web sites propose search engines to query their databases.  </p>

   <div class='protocol'>
   	<ul>
            <li>Go to <a target="_blank" href="http://www.ncbi.nlm.nih.gov/geo/">GEO web site</a>.</li>
            <li>Choose "Search" and paste GSE40129 (GSE stands for GEO Series Experiment). Click "GO" to get information about this experiment.</li>
            <li>In the "sample section" (middle of the page), click on "More" to visualize all sample names.</li>
            <li>Click on GSM986063 hyperlink (GSM stands for GEO SaMple) to get information about this sample.</li>
            <li>In the "relations" section, select "SRX176860" hyperlink to open the SRA page corresponding to this sample.</li> 
	</ul>
   </div>
   
   <div class='exo'> 
     <ul>
       <li>What is the HTS platform used to sequence this sample ?</li>
       <li>Is this experiment single-end or paire-end sequencing ? </li>
       <li>How many runs (<i>i.e.</i> lanes) are associated to this sample ? </li>
       <li>How many reads were produced (# of Spots) ?</li>
       <li>Select SRR540192 hyperlink. What is the sequence of the first read ?</li>         
       <li>What is the length of read produced ?</li>         
     </ul>
   </div>

<hr>
   <h2>Connecting to the Galaxy server </h2>


    <div class='protocol'>
      <ul>
	<li>Open a connection to pedagogix Galaxy server. </li>
	<li>Enter your login (command <i>Login</i> in the menu <i>User</i>
	  at the top of the Galaxy window). If this is your first
	  connection, use the <b>Register</b> command.</li>
      </ul>
    </div>


<hr><a name='qual'></a>
<h2>Quality control of sequencing data</h2>
<h4>Loading fastq files in galaxy</h4>

<p> Analysis of the whole dataset can be time consuming. Thus, in order to illustrate the mapping procedure, data were previously retrieved from SRA, fastq-transformed using <a href="http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software">SRA toolkit</a> (<b>fastq-dump</b> command) and mapped to the human genome (version hg19). A subset of reads that aligned onto chromosome 21 was extracted and will be used for this tutorial. Although analysis can be performed programmatically (using a shell script for instance), here, we will use the Galaxy framework. A subset of the run SRR540192 (ChIP Estrogen Receptor on MCF-7 treated with E2) is available for download (see below). The input will be processed in the later sections. </p>
      

    <div class='protocol'>
      <ul>
      <li><b>Download</b> the subset of SRR540192 run <a href="http://pedagogix-tagc.univ-mrs.fr/courses/data/ngs/siNT_ER_E2_r3_SRX176860_chr21_0.6_Noise.fastq.gz">here</a>.</li>
		<li>Open a connection to the <a target='_blank' href="http://139.124.66.4/galaxy/">Galaxy server</a>.</li>
		<li>In the upper left corner, click on <b>Unnamed history</b> and rename this workspace to 'ER_chr21_mapping'.  </li>
		<li>Select <b> DATA MANAGEMENT > Get Data > Upload File</b>. Press <b>Choose file</b> button and browse to file 'siNT_ER_E2_r3.SRR540192.chr21.andNoise.fastq'. Select fastqsanger as <b>File Format</b> and hg19 (Feb 2009) as a <b>reference genome</b>. Press <b>Execute</b>.</li>
		<li>Select <b>NGS TOOLS > NGS: QC and manipulation > FASTQ Groomer</b>. Set <b>File to groom</b> to 'siNT_ER_E2_r3.SRR540192.chr21.andNoise.fastq', and press <b>Execute</b></li>
		<li>Rename the output of the previous step into 'siNT_ER_E2_r3.chr21.groom' (use the pencil that is associated to each item of the history to edit their attributes).</li>
      </ul>
    </div>


<h3>Quality control with FastQC</h3>

<p><a href="http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc"> FastQC </a> aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis. FastQC can be run as a stand alone interactive application for the immediate analysis of small numbers of FastQ files, in a non-interactive mode (through shell commands) where it would be suitable for integrating into a larger analysis pipeline for the systematic processing of large numbers of files or through the Galaxy framework.</p>
<p>It is important to stress that although the analysis results appear to give a pass/fail result, these evaluations must be taken in the context of what you expect from your library. A 'normal' sample as far as FastQC is concerned is random and diverse. Some experiments may be expected to produce libraries which are biased in particular ways. You should treat the summary evaluations therefore as pointers to where you should concentrate your attention and understand why your library may not look random and diverse.</p>
<div class='protocol'>
      <ul>
        <li>Use <b> NGS TOOLS > NGS: QC and manipulation > FastQC:Read QC</b>.</li>
        <li>Select 'siNT_ER_E2_r3.chr21.groom' in <b>Short read data from your current history</b> dropdown list. Press <b>execute</b>.</li>
        <li> Display the data for the corresponding result in your <b>history</b> (right panel).</li>
      </ul>
</div>
<p></p>

    <div class='exo'>
      <p>
        <ul>
         <li>Carefully inspect all the statistics. What do you think of the overall quality of the sequencing ?</li>
         <li>Look carefully at the <b>Sequence Duplication Levels</b>. High level of duplicates may indicated poor heterogeneity of the fragments related to failure of the immunoprecipitation.</li>
         <li>What is the format of quality encoding. You need to know it to perform next step (read trimming).</li>
        </ul>
    </div>


<hr><a name="trim"></a>
<h2>Read trimming and filtering</h2>
<p>Read trimming is a pre-processing step in which input read ends with poor quality values are cut (most generally the right end). However one should keep in mind that this step is crucial when working with numerous aligners such as bowtie. Indeed as bowtie does not perform "hard-clipping" (that is clip sequence NOT present in the reference) it may be unable to align a large fraction of the dataset when poor quality ends are kept. Several software may be used to perform sequence trimming : </p>. 

<ul>
<li><a target="blank" href="http://hannonlab.cshl.edu/fastx_toolkit/">FASTX-Toolkit</a> (cut a defined number of nucleotides)</li>
<li><a target="blank" href="http://github.com/najoshi/sickle">sickle</a> (sliding window-based trimming)</li>
<li><a target="blank" href="http://code.google.com/p/cutadapt/">cutadapt</a> (delete ends using bwa algorithm)</li>
<li><a target="blank" href="http://bioconductor.org/packages/2.11/bioc/html/ShortRead.html">the ShortRead Bioconductor package</a> (rule may be user-defined)</li>
</ul>

<p><b>Here we will use sickle.</b></p>

<div class='protocol'>
      <ul>
        <li>Search for the <b>sickle</b> tool using the galaxy search engine (upper left corner). Select <b>sickle</b> tool.</li>
			<li>From <b>reads fastq file</b> dropdown list select'siNT_ER_E2_r3.SRR540192.chr21.andNoise.groom'. Set <b>Quality Type</b> to <b>Sanger</b> <b>Quality Threshold</b> to 20, <b>Length Threshold</b> to 25, <b>min_len</b> to 25 and press <b>execute</b>.</li>
			<li>Rename the output into siNT_ER_E2_r3.chr21.sickle</li>
			<li>Perform a new fastqc analysis using the trimmed read as input. The number of reads should be reduced.</li>
			<li>Check the proportion of duplicate reads ('Sequence Duplication Levels'). High level of PCR duplicates means that you provided to little material for sequencing (poor library complexity).</li>
      </ul>
</div>

<hr><a name="bowtie"></a>
<h2>Mapping reads with bowtie</h2>
<p>Among the genome aligners, bowtie is one of a most popular mostly because it can achieve fast alignment of millions of reads. Although, the mapping strategy differs between version 1 and 2, the overall pipeline is identical. Bowtie uses a "seed and extend" strategy meaning that it will first try to find matches for 5' ends of the reads (the seeds, whose length is controlled through -l arguments) in the reference genome (using an index build using Burrows Wheeler Transform algorithm). In the second step,  it will try to extend these matches using dynamic programming.</p>
<p>Bowtie offers many parameters that can modify the way alignment is performed. In the case of ChIP-Seq analysis, one crucial issue is to control for multi-reads (reads that map to several positions onto the reference genome) that may produce artificial peaks. This parameter may be controlled trough the<b>-m</b> arguments. Here, we will instruct bowtie to discard multihits (although more advanced policies have been proposed).</p>

<div class='protocol'>
      <ul> 
        <li>From the tool panel, select <b>NGS TOOLS > NGS: Mapping > Map with Bowtie for Illumina</b>. Fill the form as follow:.
        <ul>
	                <li>Select "Use a built-in index".</li>
        		<li>Set <b>Select a reference genome</b> to 'chr21_hg19'.</li>
        		<li>Set <b>Bowtie settings to use</b> to 'full parameter list'.</li>
        		<li>Set <b>Maximum number of mismatches permitted in the seed</b> to '1'.</li>
        		<li>Set <b>Suppress all alignments for a read if more than n reportable alignments exist (-m)</b> to 1.</li>
        		<li>Download the result from bowtie (a sam file).</li>
        </ul></li>
        </ul>
</div>



<a name="igv"></a>
<hr><h2>Viewing the results with Integrated Genome Browser (IGV).</h2>
<p> The <a href="http://www.broadinstitute.org/igv/home">Integrative Genomics Viewer</a> (IGV) is a high-performance visualization tool for interactive exploration of large, integrated genomic datasets. It supports a wide variety of data types, including array-based and next-generation sequence data, and genomic annotations.</p>


<div class='protocol'>
      <ul><li>Download IGV and launch it with 750 MB or 1.2 Gb depending of your machine.</li>
        <li>Select hg19 genome and go to chromosome 21.</li>
        <li>Select <b>Tools > Run Igvtools</b>.  Select<b>command > sort</b>, select<b>iput file</b> and browse to the sam file. Press <b>Run</b>.</li>
        <li>Select <b>command > index</b>, select<b>input file</b> and browse to the sorted sam file. Press <b>Run</b>.</li>
        <li>Select <b>command > count</b>, select<b>input file</b> and browse to the sorted sam file (! not the *.bai)). Press<b>Run</b>.</li>
        <li>Close the igvtools window.</li>
        <li>Load the tdf and sorted sam file.</li>
        <li>Go to TFF1 gene.</li>
        <li>Unzoom and and select regions displaying high signal based on tdf track</li>
</div>

<p><b>NB:</b>The tdf file is a IGV specific format that is closed to the bigWig format (the compressed version of wig format).</p>



<hr><a name="input"></a>
<h2>Mapping reads with bowtie (input)</h2>

    <div class='exo'>
      <p>Using the same procedure, align reads obtained from the input sample on hg19 genome.
      <ul>
      <li>Input reads can be obtained <a href="http://pedagogix-tagc.univ-mrs.fr/courses/data/ngs/MCF_input_r3_SRX176888_chr21_Noise.fastq.gz">here</a>.</li>
      </ul>
    </div>
    
    
<hr><a name="peaks"></a>
<h2>Peak calling</h2>

<p>Peak calling will be performed using MACS (MACS14).</p> 
<div class='protocol'>

<p>Select the tools MACS (1.4.1) in the toolbox and fill the form as below:</p>
  <ul>
    <li><b>Experiment name</b> : give a name for the MACS run.</li>
    <li><b>Paired end sequencing:</b> MACS can handle single or paired-end data; here we select single end.</li>
    <li><b>ChIP-seq tag file</b> : select a BAM file containing the treatment.</li>
    <li><b>ChIP-seq control file</b> : select the BAM file for the input</li>
    <li><b>Effective genome size</b>: 48129895.</li>
    <li><b>Tag size </b>: these are Illumina datasets of read size 36.</li>
    <li><b>Save shifted raw tag count at every bp into a wiggle file</b> : select save. This will output WIG files for the treatment and the input.</li>
    <li><b>Diagnosis report:</b>. Select <b>Produce diagnosis report</b></li>
    <li>All other options should be set to default.</li>
   </ul>
</div>

<p>MACS running using these options should generate 4 result files:</p>
<ul>
  <li>A html report describing the model built by MACS.</li>
  <li>A wig file for the treatment and control dataset.</li>
  <li>A bed file containing the peaks.</li>
</ul>

<div class='exo'>
  <p>Use the </b>sort</b> tool in the toolbox to sort MACS peaks according to the score.</p>
  <p>How many peaks have been called by MACS for each dataset ?<p>
</div>


    
<hr><a name="peaks"></a>
<h2>MACS result on the full data set</h2>
<ul>
<li> The result obtained with MACS on the full dataset (whole genome) can be obtained from: <a href="http://pedagogix-tagc.univ-mrs.fr/courses/data/ngs/MACS14_siNT_ER_E2_r1_fullGenome.bed">here</a>.</li>
</ul>
<!--
<hr><a name='troubles'>
</a><h2>If you get troubles</h2>

  <p>If you get troubles computing all steps of the analysis ...</p>

    <p> The full analysis have been pre-computed on is available has work spaces (histories) through the Galaxy web server. The following items are availables:
    <ul>
      <li><a href='http://galaxy.sb-roscoff.fr/u/dputhier/h/erchr21mappingshared'>Fastq file, Fastqc report, sam file.</a></li>
<p></p>
    <p> If needed (just do it in case there is no alternative solution), import the data using the following procedure.

        <ul>
         <li>Click on one of the link above</li>
        <li>In the top-right corner of the window, click <i>Import history</i></li>
        <li>In the message box, click the link <i>Start using this
            history</i></li>
        <li> You can manage histories using <i>options > Saved history </i>  
        </ul>
    </div>

    <hr>-->
    
 
</html>
