---
title: "Occurrence statistics: read mapping, peak calling, motif discovery and functional enrichment"
author: "Jacques van Helden and Denis Puthier"
date: '`r Sys.Date()`'
output:
  html_document:
    highlight: tango
    theme: united
    toc: yes
    toc_depth: 3
  pdf_document:
    highlight: zenburn
    toc: yes
    toc_depth: 3
css: ../../html/course.css
---


**Author:** Jacques van Helden \
**Date**: 2014-11-02

## Introduction: applications of occurrence statistics

In this series of practicals, we will analyse different data types that would *a
priori* seem to address unrelated questions, but share a common feature:
all of these problems can be treated statistics on the number of
occurrences of some biological features.

1.  [**Read mapping**](#read-mapping): number of matches expected by
    chance when mapping a library of raw reads against a genome, with a
    given number of accepted mismatches.
2.  [**Peak calling**](#peak-calling): significance of the number of
    reads found in a given genomic window, as compared to the random
    expectation estimated from the genomic input.
3.  [**Motif discovery**](#motif-discovery): detect over-represented k-mers in the peaks, in order to predict binding motifs of transcription factors putatively involved.
4.  [**Functional enrichment**](#functional-enrichment): compute the
    significance of the association between a set of genes (for example
    the genes linked to a set of ChIP-seq peaks) with a catalogue of
    functional classes (e.g. biological processes from the Gene
    Ontology).

These applications wil lead us to use some classical discrete
distributions of probabilities: hypergeometric, binomial, Poisson. For
each problem, we will have to choose the most appropriate distribution,
and to justify our choice. We will also explore the relationship between
these three distributions, and see that *under some circumstances*, one
of them can be used as a reasonable approximation of another one.

## Report format

We expect the report in the form of an [R
markdown](http://rmarkdown.rstudio.com/) document + the corresponding
pdf file.

R markdown allows to generate user-readable documents including

-   Structured text (headings, lists)
-   Inserted sections of R code
-   Results of running this R code
-   Figures generated by this R code

They can easily be generated (and exported to pdf) with the program
[Rstudio](http://www.rstudio.com/).

* * * * *

## Read mapping

We will compute the expected number of random matches when mapping a
library of *N* reads of length *k* onto a genome of size G, as a
function of the number of accepted mismatches *m*. For the sake of
simplicity, we will assume that this genome is composed of identically
and independently distributed (i.i.d.) nucleotides.

[Exercises on read
mapping](../chip-seq_read_mapping/read_mapping_statistics.html)

* * * * *

## Peak calling

After read mapping, we computed the number of reads per window of a
fixed width *L* over the whole genome, for a given ChIP-seq sample. We
also computed the number of reads per window for an control sample
(genomic input). In order to detect peaks, we need to estimate the
significance of the number of reads found in each window.

[Exercises on peak
calling](../chip-seq_peak_calling/peak_calling_statistics.html)

* * * * * 

## Motif discovery

Since ChIP-seq peaks were obtained by sequencing genomic regions bound to a particular transcription factor, we expect to find a particular enrichment for binding sites of this transcription factor. We might also discover additional motif in the peaks, that would correspond to other transcription factors interacting with the immunoprecipitated factor, and whose binding sites are frequently located at close proximity.  

In this exercise, we will run a motif discovery approach relying on the detection of over-represented k-mers in a given sequence set relative to some background model, and compare the significant k-mers with the binding motif of our factor of interest.  

* * * * *

## Functional enrichment

After peak-calling, we can map the peaks to the closest genes, and
compare the list of genes to a wide diversity of functional catalogues
(biological processes, metabolic pathways, diseases, ...). We then need
to assess the significance of the intersection between the
peak-associated genes and the genes belonging to each functional class.

[Exercises on enrichment for functional classes
(GO)](../go_statistics_td/go_statistics_td.html)

* * * * *
