---
title: "Selecting differentially expressed genes with R or TmeV"
output: html_document
---


  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <link rel="stylesheet" type="text/css" href="../../html/course.css" media="screen,projection" />
    <title></title>
    <script type="text/javascript" src="../syntaxhighlight/shCore.js"></script>
    <script type="text/javascript" src="../syntaxhighlight/shBrushBash.js"></script>
    <script type="text/javascript" src="../syntaxhighlight/shBrushR.js"></script>
    <link type="text/css" rel="stylesheet" href="../syntaxhighlight/shCore.css"/>
    <link type="text/css" rel="stylesheet" href="../syntaxhighlight/shThemeDefault.css"/>
    <script type="text/javascript">
      SyntaxHighlighter.config.clipboardSwf = '../../style/syntaxhighlight/clipboard.swf';
      SyntaxHighlighter.all();
    </script>
   
  </head>

  <body>
    <script language="JavaScript" type="text/javascript">
      <!--
          function sizeTbl2(h,i) {
          var tbl = document.getElementById(i);
          tbl.style.display = h;
          }
          // -->
    </script>


## Retrieving the den Boer normalized dataset
Here we will use the GSE13425 experiment which which was retrieved from the [Gene Expression Omnibus](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13425) (GEO) public database. In this experiment, the authors were interested in the molecular classification of acute lymphoblastic leukemia (ALL) that are characterized by the abnormal clonal proliferation, within the bone marrow, of lymphoid progenitors blocked at a precise stage of their differentiation.

Data were produced using Affymetrix geneChips (Affymetrix Human Genome U133A Array, HGU133A). Informations related to this platform are available on GEO website under identifier [GPL96](http://www.ncbi.nlm.nih.gov/projects/geo/query/acc.cgi?acc=GPL96).

<hr>

## Loading data into R


<div class="protocol">
Start R.

* Have a look at the description of the **read.table()** function.
* We will now load three data tables into R using the read.table function. The function allows us to directly read the tables from the web server. We will successively load 3 files providing complementary information.

    * the expression matrix (GSE13425_Norm_Whole.txt)
        * Contains genes as rows and samples as columns. 
        * Data were previously normalized using rma algorithm (they are thus transformed in logarithm base 2).
    * the A/P/M matrix (GSE13425_AMP_Whole.txt)
        * Indicates whether a gene was called **A**bsent, **P**resent or **M**arginal.
    * Phenotypic data (GSE13425_phenoData.txt)
        * The GSE13425_phenoData.txt file contains phenotypic information about samples.

```{r}
## Get some help about the read.table fonction
#?read.table
url.base <- "http://pedagogix-tagc.univ-mrs.fr/courses/ASG1/data/marrays/"

## Load expression values
expr.matrix <-  read.table(file.path(url.base, "GSE13425_Norm_Whole.txt"),sep="\t", head=T, row=1)

## Load phenotypic data
pheno <- read.table(file.path(url.base, 'phenoData_GSE13425.tab'), sep='\t', head=TRUE, row=1)

## Load Absent/Marginal/Present (AMP) calls
amp <- read.table(file.path(url.base, "GSE13425_AMP_Whole.txt"),sep="\t", head=T, row=1)
```

We will now define a directory to store the results. 

```{r}
## Define the output directory. You can adapt this to your local configuration.
dir.output <- "~/ASG1_practicals/GSE13425"

## Create the output directory (if it does not exist yet)
dir.create(dir.output, showWarnings=FALSE, recurs=TRUE)

## Change directory to dir.output
setwd(dir.output)
```
</div>


<div class="exo">
* How many rows and columns does the object expr.matrix contain
* Does it correspond to the dimensions of the A/P/M matrix ?
* Which information is available about samples ?
* How many samples from each tumor subtype are present in the DenBoer dataset ?
    </div>
        
    <br>
    <a href="javascript:sizeTbl2('block','1')">View solution</a>| 
    <a href="javascript:sizeTbl2('none','1')">Hide solution</a>
    <div id="1"  style="overflow:hidden;display:none">
      <h4>Solution</h4>
```{r}
## Check the dimension of the different tables
# an alternative is to use nrow and ncol
dim(expr.matrix)
dim(amp)
dim(pheno)
colnames(pheno)

## The sample title indicates the subtype of the ALL tumours. We can
## use the R function table() to count the number of samples assigned 
## to each tumour class.
as.data.frame(table(pheno$Sample.title))
```
</div>


### Interpretation

The dataset from DenBoer contains **190 samples** belonging to **various tumour classes**. We can already notice that there is an **important imbalance** between the sizes of the tumour classes: T-ALL, pre-B ALL, TEL-AML1 and hyperdiploid are each represented by more than 40 samples, whereas other classes (e.g. BCR-ABL, E2A-rearranged) are represented by a handful of samples.

The number of samples per group is a very important factor for selecting differentially expressed genes: in general, **the power of the tests** (i.e. the capacity to detect effectively differentially expressed genes) **increases with group sizes**.

<hr>
## Basics about Welch's t test
Welch's test is a variant of the classical Student test, whose goal is to test the equality between two means.
$$H_0: m_{g,1} = m_{g,2}$$

where $m_{g,1}$ and $m_{g,2}$ represent the **respective mean expression values for a given gene $g$ in two populations** (for example, all existing patients suffering from T-ALL versus all patients suffering from pre-B ALL). Of course, we do not dispose of measurements for all the patients suffering from these two types of ALL in the world (the population). We only dispose of two sets of samples, covering 36 (T-ALL) and 44 (pre-B ALL) patients, respectively. On the basis of these samples, we will estimate how likely it is that genes $g$ is generally expressed at similar levels in the populations from which the samples were drawn.

The essential difference between **Student** and **Welch** is that the proper Student test relies on the assumption that the two sampled populations have the **same variance**, whereas Welch's test is designed to treat populations with **unequal variances**.

When detecting differentially expressed genes, **we cannot assume equal variance**. Indeed, a typical case would be that a gene of interest is expressed at very low level in the first group, and high level in the second group. The inter-individual fluctuations in expression values are expected to be larger when the gene is expressed at a high level than when it is poorly expressed. It is thus generally recommended to use Welch rather than Student test when analyzing microarray expression profiles.

**BEWARE**: Student and Welch tests **assume data normality**. Affymetrix microarray intensities are far from the normal distribution, even after log transformation. However, **t-test is robust to non-normality if there is a sufficient number of samples per group**. In the subsequent exercise, we will apply Welch test to detect genes differentially expressed between cancer types represented by ~40 samples each. We are thus in **reasonably good conditions** to run a Welch test. Nevertehless, in a next section we will also apply a non-parametric test (Wilcoxon), which does not rely on an assumption of normality.

Welch's t-test defines the t statistic by the following formula:

$$ t=\frac{\bar{X_1} - \bar{X_2}}{\sqrt{\frac {s^2_1}{N_1} + \frac{s^2_2}{N_2}}}$$

Where:

* $\bar{X_{i}}$ is the sample mean,
* $s^2_{i}$ the sample variance,
* $N_{i}$ the sample size.

The **t.test()** function can be used to calculate this score (and additional informations such as p.value). This function returns an **S3 object** whose slots can be listed using the **names()** function and accessed using the **$ operator** (such as with lists in R).

### A first intuition
In order to get an intuition of the t statistics, let us create artificial datasets and compute the associated t value. In the following example x and y can be viewed as the values for gene g in class x and y.

```{r}
## Generate 4 random numbers following a normal distribution, to
## simulate the samples 1 and 2. We deliberately set the means to 
## the same values (to fall under the null hypothesis), but 
## different standard deviations.
x <- rnorm(n=4, mean=6, s=1)
y <- rnorm(n=4, mean=6, s=2)
```

<div class="exo">
* Compute the associated $t$ value using the **mean**, **sd** and **sqrt** functions.
</div>
        
<br>
<a href="javascript:sizeTbl2('block','2')">View solution</a>| 
<a href="javascript:sizeTbl2('none','2')">Hide solution</a>
<div id="2"  style="overflow:hidden;display:none">
<h4>Solution</h4>
```{r}
## Compute the t statistics manually
nx <- length(x)
ny <- length(y)
diff <- mean(x) - mean(y)
t.obs <- diff/sqrt((sd(x)^2)/nx + (sd(y)^2)/ny)

## print the result
print(t.obs) # or t.obs or show(t.obs)
```
</div>

<div class="exo">
* Now we can check that the same result is obtained using the **t.test** function implemented in R.
</div>

<br>
<a href="javascript:sizeTbl2('block','3')">View solution</a>| 
<a href="javascript:sizeTbl2('none','3')">Hide solution</a>
<div id="3"  style="overflow:hidden;display:none">
<h4>Solution</h4>
```{r}
## Run the Welch test (this is specified by indicating that we don't expect equal variances)
simulated.welch <- t.test(x,y, var.equal=FALSE)
print(simulated.welch)
## note: each student should have a different result, since values were generated at random

## Retrieve the t statistics
names(simulated.welch)
simulated.welch$statistic

## Compare the t statistics computed by the t.test() function and your manual computation
simulated.welch$statistic == t.obs
```
</div>


<hr>

## Applying Welch's t-test to the den Boer dataset

We would like to define genes **that discriminate between "TEL-AML1" tumors and "hyperdiploid" tumors**.

One possibility would be to iterate over all probesets, and to successively run the R method t.test() on each one. This would however be quite inefficient, and the results would not be very easy to handle, since it would be a list of objects of the class t.test.

Instead, we will use a custom function that runs Student or Welch test in parallel on all the elements of a data table.

### Running t-tests on each row of a data matrix

<div class="protocol">
#### Installing the qvalue library
First we need to check if the qvalue library is installed (we will give more information about q-values in the next sessions).
```{r}
### Running t-tests on each row of a data matrix
## We must first check if the q-value library from Bioconductor has
## been installed (if not, will be installed here)
source("http://bioconductor.org/biocLite.R")
if (!require("qvalue")) { biocLite("qvalue") }
```

#### Loading the t.test.multi function
The we will load a custom script written by J. van Helden.

```{r}
## Load a custom the library for multiple t tests
source('http://www.bigre.ulb.ac.be/courses/statistics_bioinformatics/R-files/config.R')
source('http://www.bigre.ulb.ac.be/courses/statistics_bioinformatics/R-files/util/util_student_test_multi.R')
```

For the sake of curiosity, you can also have a look at the [R code](http://www.bigre.ulb.ac.be/courses/statistics_bioinformatics/R-files/util/util_student_test_multi.R).
</div>

#### Compute t-test for all genes
We will now run the **Welch test** on **all the genes** of the Den Boer dataset. We will apply the test to select genes differentially expressed between the subtypes "hyperdiploid" and "TEL-AML1".

```{r}
## Define a Boolean vector indicating which samples belong to the two selected subtypes.
samples.to.keep <- pheno$Sample.title == "hyperdiploid" |  pheno$Sample.title == "TEL-AML1"

## Define a vector with the sample types for the two selected cancer subtype
cancer.type <- as.vector(pheno[samples.to.keep, "Sample.title"])
print(cancer.type)
table(cancer.type)

## Run the Welch test on each probesets of the DenBoer expression matrix
denboer.welch <- t.test.multi(expr.matrix[samples.to.keep], cancer.type, volcano.plot=FALSE)

## Inspect the result table
dim(denboer.welch)
names(denboer.welch)

## Select genes with a stringent threshold on E-value
threshold <- 0.05
significant.probesets <- denboer.welch$E.value <= threshold
table(significant.probesets) ## Count the number of significant probesets
```


#### Comparing sample means

We will compare the mean expression value between the two subtypes, and highlight the significant genes.


```{r}
## Plot the gene-wise means
plot(denboer.welch[, c("mean.TEL-AML1", "mean.hyperdiploid")], col="darkgray")
grid()
abline(a=0,b=1, col="black")
lines(denboer.welch[significant.probesets, c("mean.TEL-AML1", "mean.hyperdiploid")], type='p', col="red") ## Highlight significant genes
```


<div class="exo">
* How do you explain that the regions covered by gray (non-significant) and red (significant) probesets overlap on the mean-mean plot ?
</div>

<br>
<a href="javascript:sizeTbl2('block','4')">View solution</a>| 
<a href="javascript:sizeTbl2('none','4')">Hide solution</a>
<div id="4"  style="overflow:hidden;display:none">
<h4>Solution</h4>
The significance of a Welch (or a Student) test depends not only on the differences between the means, but also on the estimation of the standard deviation of this difference. In other terms, a same difference (or a same ratio) between two means could be either significant or not, depending on whether the two groups to be compared have a high or low variance.
</div>

<hr>

## Comparing the p-values of Welch and Wilcoxon tests

### The apply function
The **apply** function can be used to apply a given function to a matrix or data.frame. This function has tree required arguments:

```{r}
args(apply)
```
* X the matrix/data.frame
* MARGIN: 1 or 2 depending on wether the function is to be applied on rows or columns respectively. 

### Defining a new function: return.t
In the line below, we define a function called return.t, that can be used to run the Welch test on a single probeset of the microarray table.

```{r}
## Define a function to return the p-value of a Welch test
return.t <- function(x,y){  t.test(x[y=="TEL-AML1"], x[y=="hyperdiploid"], alternative="two.sided", var.equal=FALSE, paired=FALSE)$p.value}
```

<div class="exo">

* Use this function to compute the p-value of the Welch's t test for all probesets of expr.matrix.
* Define a similar function to compute the p-value of Wilcoxon's test to each probeset.
* Draw a plot to compare the p-values returned by the respective tests.
</div>

<br>
<a href="javascript:sizeTbl2('block','5')">View solution</a>| 
<a href="javascript:sizeTbl2('none','5')">Hide solution</a>
<div id="5"  style="overflow:hidden;display:none">
<h4>Solution</h4>
```{r}
## Define a vector containing sample name (classe information) 
cl <- pheno$Sample.title

## Define a function to return the p-value of a Wilcoxon test
return.wilcox <- function(x,y){ wilcox.test(x[y=="TEL-AML1"], x[y=="hyperdiploid"], alternative="two.sided", paired=FALSE, exact=FALSE)$p.value }

## Create a data frame with the p-values of the Welch and Wilcoxon tests
diff.tests <- data.frame(
        welch.pval=apply(expr.matrix,1,return.t,cl),
	      wilcox.pval=apply(expr.matrix,1,return.wilcox,cl)
	      )

pval.threshold <- 1e-3


## Compute a contigency table counting the number of consistent / different results between Welch and Wilcoxon tests
table(diff.tests$welch.pval < pval.threshold, diff.tests$wilcox.pval < pval.threshold)

################################################################
## Plot the respective p-values returned by the two tests

pch <- "."
cex <- 3
min.pval <- min(as.vector(as.matrix(diff.tests[, c("welch.pval", "wilcox.pval")])))
plot(diff.tests$welch.pval, diff.tests$wilcox.pval, log="xy", panel.first=grid(),  xlim=c(min.pval, 1),  ylim=c(min.pval, 1), col="gray", xlab="Welch test p-value", ylab="Wilcoxon test p-value", main="DEG selection in Den Boer (2009), TEL-AML1 vs hyperdiploidy", pch=pch, cex=cex)

## Highlight in green the genes selected by both methods
welch.and.wilcox <- diff.tests$welch.pval < pval.threshold & diff.tests$wilcox.pval < pval.threshold
points(diff.tests[welch.and.wilcox,c("welch.pval", "wilcox.pval")], col="darkgreen", pch=pch, cex=cex)

## Highlight probesets whose selection is affected by the choice of the test
wilcox.not.welch <- diff.tests$welch.pval >= pval.threshold & diff.tests$wilcox.pval < pval.threshold
points(diff.tests[wilcox.not.welch,c("welch.pval", "wilcox.pval")], col="red", pch=pch, cex=cex)

welch.not.wilcox <- diff.tests$welch.pval < pval.threshold & diff.tests$wilcox.pval >= pval.threshold
points(diff.tests[welch.not.wilcox,c("welch.pval", "wilcox.pval")], col="orange", pch=pch, cex=cex)


## Draw lines to display the thresholds on the respective tests
abline(v=pval.threshold,col="red", lwd=2)
abline(h=pval.threshold,col="red", lwd=2)
abline(h=1e-16,col="gray") ## Draw the limit of floating point calculation, which is the limit for p.value in wilcox.test()
abline(v=1e-16,col="gray")
abline(a=0,b=1, col="black", lwd=2)

```
</div>

<hr>
## Drawing a volcano plot

The volcano plot is a classical representation of differential expression analysis. In this diagram, the **x axis represents the log** ratio and the **y axis the result of a statistic** expressed as $-log10(p-value)$.

### Computing the log ratio

<div class="exo">
* Calculate for each gene its average expression level in "TEL-AML1" and "hyperdiploid" class.
* Calculate the difference of the mean for each gene (log ratio).
</div>

<br>
<a href="javascript:sizeTbl2('block','6')">View solution</a>| 
<a href="javascript:sizeTbl2('none','6')">Hide solution</a>
<div id="6"  style="overflow:hidden;display:none">
<h4>Solution</h4>
```{r}
rowMeans.tel <- rowMeans(expr.matrix[,cl== "TEL-AML1"])
rowMeans.hyp <- rowMeans(expr.matrix[,cl== "hyperdiploid"])
diff <- rowMeans.tel - rowMeans.hyp
```
</div>

### Volcano plot

<div class="exo">
* Draw a volcano plot.
* Use the identify function to the names of some interesting genes.
</div>

<br>
<a href="javascript:sizeTbl2('block','7')">View solution</a>| 
<a href="javascript:sizeTbl2('none','7')">Hide solution</a>
<div id="7"  style="overflow:hidden;display:none">
<h4>Solution</h4>
```{r}
## Compute the significance, i.e. -log10 of the p-value
t.res <- diff.tests$welch.pval
mlt <- -log10(t.res)

## Draw the Volcano plot
plot(diff,mlt,pch=16,cex=0.7,xlab="Log ratio (base 2)",ylab="log10(1/p-value)")
grid()

## Select probesets based on two criteria (fold change + p-value)
retained <- (abs(diff) > 1) & (t.res < 1e-3)

## Color the selected probesets
points(diff[retained],mlt[retained],col="red",cex=0.7,pch=16)

## Draw the selection thresholds
abline(v=c(-1,1), col="violet")
abline(h=3, col="violet")
```
</div>

<hr>
## Significance Analysis of Microarrays (SAM)
### What is SAM
The most popular method for differential expression analysis of microarray data is "Significance Analysis of Microarrays" (SAM). SAM will compute for each gene a score $d$ which is close to the $t$ statistics of the welch's test. However, it won't require any assumption about the data distribution.  In order to compute the expected distribution of $d$ under the null hypothesis SAM will performe a set of permutations on the class label and compute each time a simulated sets of results for $d$ statistics. The observed results and simulated results will be used to compute the FDR values.
Sam is implemented in several R library (e.g: siggenes). Here we will use a more interactive program called ["MultiExperiment Viewer (MeV)](http://www.tm4.org/mev.html).

### Installing Mev on Linux system
If you are working on a Linux system, the commands below can be used to download and compule MeV on the Linux console.

<pre class="brush:bash">
mkdir -p ~/bin
cd ~/bin
wget "http://freefr.dl.sourceforge.net/project/mev-tm4/mev-tm4/MeV%204.8.1/MeV_4_8_1_r2727_linux.tar.gz"
tar xvfz MeV_4_8_1_r2727_linux.tar.gz
cd -
echo -e "\nalias tmev=\"cd ~/bin/MeV_4_8_1/; sh -c 'nohup ./tmev.sh &'; cd -\"" >> ~/.bashrc
source ~/.bashrc
tmev
</pre>
  
### Filtering genes on the basis of the absent/marginal/present (A/M/P) filter
The classical processing pipeline defined by Affymetrix associates a qualitative tag to each probeset, with three possible values:

* absent (A)
* marginal (M)
* present (P)
An **absent** call means that no significant signal was detected with the associated probeset. However, this "absence" might either indicate that the gene is not expressed in this particular sample, or that the gene is undetectable (irrespective of its expression level) due to some technical problem with this specific probeset.

It became a classical practice to filter out the genes called "absent" on an important fraction of the samples in a given series, by implicitly assuming that their recurrent absence reveals a technical problem rather than a biologically relevant effect (repression of the gene).

In the following exercise, we will apply an A/M/P filter to discard the genes declared absent in at least 30% of the genes. Alternatively, if you want to skip this exercice you can directly get the output file GSE13425_Norm_TEL-AML1vsHyperdip.txt [here](~/courses/data/microarrays/GSE13425_Norm_TEL-AML1vsHyperdip.txt).

<div class="exo">
Select a subset of the dataset (expr.matrix.sub) that will contain only "TEL-AML1" tumors and "hyperdiploid" tumors.
Create a vector cl that will contain information regarding the classes of samples stored in expr.matrix.sub ("hyperdiploid" or "TEL-AML1").
Select genes giving a signal ("present" call) in at least 30% of the selected samples.
</div>

<br>
<a href="javascript:sizeTbl2('block','8')">View solution</a>| 
<a href="javascript:sizeTbl2('none','8')">Hide solution</a>
<div id="8"  style="overflow:hidden;display:none">
<h4>Solution</h4>
```{r}
## Select the samples from the two selected ALL classes
samples.to.keep <- pheno$Sample.title == "hyperdiploid" |  pheno$Sample.title == "TEL-AML1"

## Check samples to keep
print(samples.to.keep)
sum(samples.to.keep)
print(as.vector(pheno[samples.to.keep, "Sample.title"]))

## Extact a subset of expression matrix with only the two selected sets
expr.matrix.sub <- expr.matrix[,samples.to.keep]

## Define a vector indicating sample groups
cl <- as.vector(pheno[samples.to.keep,]$Sample.title)
print(cl)

## Select a subset of the A/M/P matrix
amp.sub <- amp[,samples.to.keep]
dim(amp)
dim(amp.sub)

## Count the number of "Present" calls per probeset
isPresent <- amp.sub == "P"
present.per.probeset <- rowSums(isPresent)
hist(present.per.probeset, breaks=0:ncol(amp.sub))

## "Present filter":  Select probeset declared present in at least 25% of the samples
retained <- rowSums(isPresent) >=  0.25*ncol(expr.matrix.sub)
table(retained) ## Count number of retained and rejected probesets

## Select a subset of the matrix with the retained probesets only
expr.matrix.sub <- expr.matrix.sub[retained, ]

## Check the number of probes (rows) and samples (columns) of the 
## selected expression table
print(dim(expr.matrix.sub))

## Export the table
setwd(dir.output)
write.table(expr.matrix.sub, "GSE13425_Norm_TEL-AML1vsHyperdip.txt", col.names=NA,quote=F,sep="\t")
```

### Applying SAM algorithm with MeV

<div class="protocol">
* Load the file using "File > Load data > Select file loader Tab delimited".
* **Browse** to file "GSE13425_Norm_TEL-AML1vsHyperdip.txt", click on the **upper-leftmost expression value** and click on the
	  **load** button.
* Select **Adjust data > Gene/Rows adjustment > median center Genes/Rows**
* Select **Analysis > Statistics > SAM**
* Set all samples from GSM338746.CEL.gz to GSM338789.CEL.gz to class B.
* Set the number of permutations to 500, select **Construct hierachical clustering** and click " OK".
* Accept default parameters for hierarchical clustering.
* Set the **delta value to 2** and click OK.
* Select Analysis **results > SAM > Hierarchical trees > All Significant genes**.
* Select **Display > Set color scale limits** and **set lower limit to -4**, **midpoint value to 0** and **upper limit to 4**.
* Set **Display > Set Element Size** to **5x2**.
* To store the resulting file right click to select the whole gene tree and select **Save cluster**.
</div>

